### PR TITLE
55 - MySQL 5.7 client compatibility patch

### PR DESCRIPTION
**Pull request by divaykin**
Add full set of required / optional fields & request native_password authentication method.

### CODE REVIEW

**Comment by divaykin** [comment]
before:
```
Jan 08 15:35:18 u1604 bedrock[3895]: xxxxx (STCPServer.cpp:57) operator() [main] [dbug] Accepting socket from '127.0.0.1:38230' on port 'localhost:3306'
Jan 08 15:35:18 u1604 bedrock[3895]: xxxxx (BedrockServer.cpp:473) postSelect [main] [info] Plugin 'MySQL' accepted a socket from '127.0.0.1:38230'
Jan 08 15:35:18 u1604 bedrock[3895]: xxxxx (MySQL.cpp:252) onPortAccept [main] [info] {MySQL} Accepted MySQL request from '127.0.0.1:38230'
Jan 08 15:35:18 u1604 bedrock[3895]: xxxxx (STCPManager.cpp:152) postSelect [main] [dbug] Connection to '127.0.0.1:38230' died (recv=0, send=1)
Jan 08 15:35:18 u1604 bedrock[3895]: xxxxx (STCPManager.cpp:255) closeSocket [main] [dbug] Closing socket '127.0.0.1:38230'
Jan 08 15:36:24 u1604 bedrock[3895]: xxxxx (STCPServer.cpp:57) operator() [main] [dbug] Accepting socket from '127.0.0.1:38232' on port 'localhost:3306'
Jan 08 15:36:24 u1604 bedrock[3895]: xxxxx (BedrockServer.cpp:473) postSelect [main] [info] Plugin 'MySQL' accepted a socket from '127.0.0.1:38232'
Jan 08 15:36:24 u1604 bedrock[3895]: xxxxx (MySQL.cpp:252) onPortAccept [main] [info] {MySQL} Accepted MySQL request from '127.0.0.1:38232'
Jan 08 15:36:24 u1604 bedrock[3895]: xxxxx (STCPManager.cpp:152) postSelect [main] [dbug] Connection to '127.0.0.1:38232' died (recv=0, send=1)
Jan 08 15:36:24 u1604 bedrock[3895]: xxxxx (STCPManager.cpp:255) closeSocket [main] [dbug] Closing socket '127.0.0.1:38232'

denis@u1604:~$ mysql --host 127.0.0.1
ERROR 2027 (HY000): Malformed packet
```

after:
```
Jan 8 21:30:49 u1604 bedrock: xxxxx (BedrockServer.cpp:525) postSelect [main] [info] Plugin 'MySQL' accepted a socket from '127.0.0.1:38298'
Jan 8 21:30:49 u1604 bedrock: xxxxx (MySQL.cpp:189) onPortAccept [main] [info] {MySQL} Accepted MySQL request from '127.0.0.1:38298'
Jan 8 21:30:49 u1604 bedrock: xxxxx (MySQL.cpp:199) onPortRecv [main] [dbug] {MySQL} Received command #-123: '2200000185A400000064656E697300006D7973716C5F6E61746976655F70617373776F726400'
Jan 8 21:30:49 u1604 bedrock: xxxxx (MySQL.cpp:274) onPortRecv [main] [info] {MySQL} Sending OK
Jan 8 21:30:49 u1604 bedrock: xxxxx (MySQL.cpp:199) onPortRecv [main] [dbug] {MySQL} Received command #3: '210000000373656C65637420404076657273696F6E5F636F6D6D656E74206C696D69742031'
Jan 8 21:30:49 u1604 bedrock: xxxxx (MySQL.cpp:205) onPortRecv [main] [info] {MySQL} Processing query 'select @@version_comment limit 1'
Jan 8 21:30:49 u1604 bedrock: xxxxx (MySQL.cpp:217) onPortRecv [main] [info] {MySQL} Returning variable 'version_comment'='30c67b648f7ce4862379c235934a7abee880b801'
Jan 8 21:30:59 u1604 bedrock: xxxxx (STCPManager.cpp:152) postSelect [main] [dbug] Connection to '127.0.0.1:38298' died (recv=0, send=1)
Jan 8 21:30:59 u1604 bedrock: xxxxx (STCPManager.cpp:255) closeSocket [main] [dbug] Closing socket '127.0.0.1:38298'

denis@u1604:~/dev/bedrock/Bedrock/plugins$ mysql --host 127.0.0.1
Welcome to the MySQL monitor. Commands end with ; or \g.
Your MySQL connection id is 1
Server version: bedrock 30c67b648f7ce4862379c235934a7abee880b801

Copyright (c) 2000, 2016, Oracle and/or its affiliates. All rights reserved.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql> ^DBye
denis@u1604:~/dev/bedrock/Bedrock/plugins$ mysql --version
mysql Ver 14.14 Distrib 5.7.16, for Linux (x86_64) using EditLine wrapper
```

**Comment by quinthar** [comment]
FYI, this PR is fixing: https://github.com/Expensify/Bedrock/issues/55

File: plugins/MySQL.cpp [hunk 1]
```diff
@@ -81,6 +81,33 @@ string MySQLPacket::serializeHandshake() {
...
+
+    const unsigned char latin1_swedish_ci = 0x08;
+    SAppend(handshake.payload, &latin1_swedish_ci, 1); // character_set
+
+    const unsigned short SERVER_STATUS_AUTOCOMMIT = 0x0002;
+    SAppend(handshake.payload, &SERVER_STATUS_AUTOCOMMIT, 2); // status_flags
+
+    SAppend(handshake.payload, &capability_flags_2, 2); // capability_flags_2 (high 2 bytes)
+
+    const unsigned char auth_plugin_data[] = {
```
Comment (quinthar) [inline plugins/MySQL.cpp]: Where did you get these values, and what do they mean?

File: plugins/MySQL.cpp [hunk 2]
```diff
@@ -81,6 +81,33 @@ string MySQLPacket::serializeHandshake() {
     SAppend(handshake.payload, &connectionID, 4); // connection_id
     handshake.payload += (string) "xxxxxxxx";     // auth_plugin_data_part_1
     handshake.payload += lenEncInt(0);            // filler
+
+    static const unsigned CLIENT_LONG_PASSWORD = 0x00000001;
```
Comment (quinthar) [inline plugins/MySQL.cpp]: The codebase doesn't use `static const unsigned` anywhere else.  Don't use `static` or `const` unless strictly needed (which isn't the case here), and use `uint32_t` as done elsewhere in this file.  Basically, please make this code look like the other code in the codebase.

File: plugins/MySQL.cpp [hunk 3]
```diff
@@ -81,6 +81,33 @@ string MySQLPacket::serializeHandshake() {
...
+    static const unsigned CLIENT_PLUGIN_AUTH   = 0x00080000;
+    const unsigned capability_flags = CLIENT_LONG_PASSWORD | CLIENT_PLUGIN_AUTH;
+
+    const unsigned short capability_flags_1 = (const unsigned short)(capability_flags);
+    const unsigned short capability_flags_2 = (const unsigned short)(capability_flags >> 16);
+    SAppend(handshake.payload, &capability_flags_1, 2); // capability_flags_1 (low 2 bytes)
```
Comment (quinthar) [inline plugins/MySQL.cpp]: It looks like you are hand-rolling the encoding, which works, but is inconsistent with the rest of the file.  Please use `lenEncInt()` to be consistent with the rest of the file.

Comment (divaykin) [inline plugins/MySQL.cpp]: @quinthar I don't really understand how `lenEncInt()` is supposed to work for a fixed length integer. Say protocol requires 2 bytes int of `capability_flags` and these flags just happens to have a value < 256. `lenEncInt()` as far as I understand would just encode it as one byte which would violate the protocol?

**Pull request merged**